### PR TITLE
Docs build improvements

### PR DESCRIPTION
**Pull request by wadells**
### Summary

This PR introduces a new make target `docs-lint` which runs `milv -ignore-external` on all of our docs. As well, the regular `make docs` and `make run-docs` now use the `--strict` flag to catch issues.  Both of these changes are building towards enforcement of these targets on all pull requests.

Why?  We had tens to hundreds of broken **_internal_** links in our customer facing documentations that @benarent and I cleaned up in #1210. Providing tooling to catch link rot and enforcing that will prevent issues like this from creeping back in.  This also establishes an engineering norm that docs quality is enforced by automatic tools.

In addition to the necessary changes, I also refactored our docs Makefile substantially, to better use partial builds and caching.  See individual commit messages for more information about those changes.

Contributes to fixing #1220.

### Testing Done

`make docs` and `make docs-lint` on this branch before rebase:
[docs-build-dirty.txt](https://github.com/gravitational/gravity/files/4315484/docs-build-dirty.txt)

`make docs` and `make docs-lint` on this branch after rebase (updated):
[docs-build-clean.txt](https://github.com/gravitational/gravity/files/4320063/docs-build-clean.txt)

I also have a PR job basically ready to run these new targets against every PR to gravity.  See https://jenkins.gravitational.io/job/gravity-pr-docs/ and the test PRs here:

https://github.com/wadells/gravity/pull/1
https://github.com/wadells/gravity/pull/2
https://github.com/wadells/gravity/pull/3

We just need to flip it over from my personal repo to the main repo and configure webhook pushes to make it go live.

### Removal of sass

@alex-kovoy @benarent: I've removed sass from our docker file in 7528d552. This should be safe @alex-kovoy removed the last scss/sass files in 2e1724ed7e3c7742c7c32a114f028108f77bda7d:

```
walt@work:~/git/gravity/docs$ git log --all --oneline -- "*.sass"                                                                                                                             
walt@work:~/git/gravity/docs$ git log --all --oneline -- "*.scss"                                                                                                                             
2e1724ed (docs) cleanup                                                                                                                                                                       
b580aeac Wolfe/docs (#79)                                                                                                                                                                     
ff8241d9 (tag: 5.2.0-alpha.2) Move all OSS code to this repository.                                                                                                                           
45b74e98 (tag: 5.2.0-alpha.0) Add OSS code.                                                                                                                                                   
walt@work:~/git/gravity/docs$ git log --stat -1 2e1724ed                                                                                                                                      
                                                                                                                                                                                              
commit 2e1724ed7e3c7742c7c32a114f028108f77bda7d
Author: Alexey Kontsevoy <alex@kontsevoy.com>
Date:   Fri Nov 2 16:18:42 2018 -0400

    (docs) cleanup

 docs/build.sh                              |    10 +-
 docs/theme/css/highlight.css               |   125 --
 docs/theme/css/telekube-bundle.css         |    46 -
 docs/theme/css/theme.css                   |    12 -
 docs/theme/head.html                       |     3 +-
 docs/theme/js/base.js                      |    76 -
 docs/theme/js/grvlib.js                    |   619 --------
 docs/theme/js/highlight.pack.js            |     2 -
 docs/theme/js/jquery-2.1.1.min.js          |     4 -
 docs/theme/js/modernizr-2.8.3.min.js       |     1 -
 docs/theme/js/theme.js                     |   262 ----
 docs/theme/scripts.html                    |    16 +-
 docs/theme/{js/grvlib.css => src/docs.css} |    26 +
 docs/theme/src/docs.js                     | 11189 +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 docs/theme/src/index.scss                  |    94 --
 docs/theme/{js => src}/prettify-yaml.js    |     0
 docs/theme/{js => src}/prettify.js         |     0
 docs/theme/{js => src}/third-party.js      |     0
 18 files changed, 11226 insertions(+), 1259 deletions(-)
```

### CODE REVIEW

File: docs/Makefile [hunk 1]
```diff
@@ -1,41 +1,91 @@
...
-# builds a docker container which is used for running `mkdocs`
-.PHONY:bbox
-bbox:
+.PHONY:docbox
+$(DOCBOX_ID_FILE): Dockerfile
+	mkdir -p $(BUILDDIR)
 	docker build \
 		--build-arg UID=$$(id -u) \
 		--build-arg GID=$$(id -g) \
 		--build-arg HOME=$(HOME) \
```
Comment (wadells) [inline docs/Makefile]: Does this need the WORKDIR change from above?  It doesn't seem to matter to the end user.

File: docs/Makefile [hunk 2]
```diff
@@ -1,41 +1,91 @@
...
+LINTS=$(addprefix lint-, $(VERSIONS))
+
+.PHONY:container-lint
+container-lint: $(LINTS)
+
+$(LINTS): lint-%:
+	cd $(*F) && milv -ignore-external
+
+.PHONY:$(LATEST_SYMLINK)
+$(LATEST_SYMLINK): $(LATEST_BUILD) $(BUILDDIR)
```
Comment (wadells) [inline docs/Makefile]: Neither of these dependencies makes sense.  I believe they're vestigial from some work discarded along the way.

File: docs/Dockerfile [hunk 3]
```diff
@@ -13,18 +13,15 @@ RUN groupadd jenkins --gid=$GID -o && useradd jenkins --uid=$UID --gid=$GID --cr
...
-RUN (apt-key update && \
-             apt-get -q -y update --fix-missing && \
-             apt-get -q -y update && \
-             apt-get install -q -y ruby-sass)
-             
-ENV LANGUAGE="en_US.UTF-8" \
-    LANG="en_US.UTF-8" \
-    LC_ALL="en_US.UTF-8" \
-    LC_CTYPE="en_US.UTF-8" \
+RUN apt-get update -y && apt-get install -q -y ruby-sass parallel
```
Comment (wadells) [inline docs/Dockerfile]: parallel isn't needed anymore, with the death of build.sh

File: Makefile [hunk 4]
```diff
@@ -746,6 +746,11 @@ docs:
 run-docs:
 	$(MAKE) -C docs run
 
+.PHONY: docs-lint
+docs-lint:
+	$(MAKE) -C docs lint
+
+
```
Comment (wadells) [inline Makefile]: extra whitespace

**Comment by wadells** [comment]
Macro question: Should the `docs-lint` target be separate from the `docs` target?  Or should `docs` depend on `docs-lint` to make this easier to use.

File: Makefile (hunk 4 above)
Comment (a-palchikov) [reply Makefile]: I have to ask - is this a question or statement? Why not remove it?

File: docs/Dockerfile (hunk 3 above)
Comment (a-palchikov) [reply docs/Dockerfile]: Same here - if it's not needed anymore - remove it.

File: docs/Dockerfile [hunk 5]
```diff
@@ -13,18 +13,15 @@ RUN groupadd jenkins --gid=$GID -o && useradd jenkins --uid=$UID --gid=$GID --cr
...
+
+# Install Go & milv, go version doesn't particularly matter as it isn't used in the product
+RUN mkdir -p /opt /gopath
+ENV GOROOT="/opt/go" \
     GOPATH="/gopath" \
-    GOROOT="/opt/go" \
     PATH="$PATH:/opt/go/bin:/gopath/bin"
+RUN cd /opt && curl https://dl.google.com/go/go1.14.linux-amd64.tar.gz | tar xz
+RUN GO111MODULE=on go get -u -v github.com/magicmatatjahu/milv@v0.0.6
```
Comment (a-palchikov) [inline docs/Dockerfile]: How about move all things only relevant for building the prerequisites (like Go for building milv here) to the build container (perhaps as a multistage build) and have the docs container without it?

File: docs/Makefile [hunk 6]
```diff
@@ -1,41 +1,91 @@
...
+	$(DOCKERMAKE) container-run RUN_CFG=$(RUN_CFG)
 
-# install mkdocs on this machine 
-.PHONY:mkdocs
-mkdocs:
-	pip install -q -U mkdocs
+# drops you into bash shell of mkdocs container
+.PHONY:shell
+shell: $(DOCBOX_ID_FILE)
+	docker run -ti $(NOROOT) $(DOCKERFLAGS) $(DOCKERPORTS) $(DOCBOX) \
```
Comment (a-palchikov) [inline docs/Makefile]: Also `--rm`?

File: docs/Makefile (hunk 2 above)
Comment (a-palchikov) [reply docs/Makefile]: Let's remove it then.

File: docs/Makefile [hunk 7]
```diff
@@ -1,41 +1,91 @@
...
+		/bin/bash
 
-# builds a docker container which is used for running `mkdocs`
-.PHONY:bbox
-bbox:
+.PHONY:docbox
```
Comment (a-palchikov) [inline docs/Makefile]: What's `docbox`?

**Review by a-palchikov — APPROVED** [review]

File: Makefile (hunk 4 above)
Comment (wadells) [reply Makefile]: I will remove it.  I only saw this issue when I was putting the PR up at the end of the day.

File: docs/Dockerfile (hunk 5 above)
Comment (wadells) [reply docs/Dockerfile]: Good idea.  I'll look into this.

File: docs/Makefile (hunk 7 above)
Comment (wadells) [reply docs/Makefile]: Good catch.  I don't remember why this changed.  It should be returned to be consistent.

File: docs/Makefile (hunk 6 above)
Comment (wadells) [reply docs/Makefile]: `--rm=true` is in the definition of `DOCKERFLAGS` on line 11.

File: docs/Makefile (hunk 2 above)
Comment (wadells) [reply docs/Makefile]: removed in b4b64f6

File: Makefile (hunk 4 above)
Comment (wadells) [reply Makefile]: removed in b1a2ffd

File: docs/Dockerfile (hunk 3 above)
Comment (wadells) [reply docs/Dockerfile]: removed in b4b64f6.  I was able to go a bit further and remove sass (and thus all the apt directives) in 7528d55.

File: docs/Dockerfile (hunk 5 above)
Comment (wadells) [reply docs/Dockerfile]: Great suggestion.  The Dockerfile is considerably simpler after I implemented this change in 
9c8fd4a.

File: docs/Makefile (hunk 1 above)
Comment (wadells) [reply docs/Makefile]: fixed in ee1376f

File: docs/Makefile (hunk 7 above)
Comment (wadells) [reply docs/Makefile]: removed in b4b64f6

**Review by alex-kovoy — APPROVED** [review]

**Review by benarent — APPROVED** [review]

**Pull request merged**